### PR TITLE
feat: Adding ProgressRing extension

### PR DESCRIPTION
**Pull request by nickrandolph**
Fixes: https://github.com/unoplatform/uno.chefs/issues/835

## PR Type

What kind of change does this PR introduce?
- Bugfix
- Feature

## What is the current behavior?

ProgressRing contained within the loading template doesn't ever stop (ie IsActive is true)

## What is the new behavior?

New ProgressRingExtensions.IsActive can be used to start/stop all nested ProgressRings
In VisualState for Loading/Loaded any ProgressRings can have IsActive set to true/false

## PR Checklist

Please check if your PR fulfills the following requirements:
- [ ] Tested code with current [supported SDKs](../README.md#supported)
- [ ] Tested the changes where applicable:
	- [ ] UWP
	- [ ] WinUI
	- [ ] iOS
	- [ ] Android
	- [ ] WASM
	- [ ] MacOS
- [ ] Updated the documentation as needed:
	- [ ] [General Doc Update](https://github.com/unoplatform/uno.toolkit.ui/tree/main/doc)
	- [ ] [Controls Doc Update](https://github.com/unoplatform/uno.toolkit.ui/tree/main/doc/controls)
	- [ ] [Extensions Doc Update](https://github.com/unoplatform/uno.toolkit.ui/tree/main/doc/helpers)
	- [ ] [controls-styles.md](https://github.com/unoplatform/uno.toolkit.ui/blob/main/doc/controls-styles.md)
	- [ ] [lightweight-styling.md (LightWeight Styling Resource Keys)](https://github.com/unoplatform/uno.toolkit.ui/blob/main/doc/lightweight-styling.md)
- [ ] [Runtime Tests and/or UI Tests](https://platform.uno/docs/articles/contributing/guidelines/creating-tests.html) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal)
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Review by kazo0 — APPROVED** [review]

File: samples/Uno.Toolkit.Samples/Uno.Toolkit.Samples.Shared/Content/Controls/LoadingViewSample.xaml
Comment (Xiaoy312) [inline samples/Uno.Toolkit.Samples/Uno.Toolkit.Samples.Shared/Content/Controls/LoadingViewSample.xaml:1]: avoid changing the xaml style

File: src/Uno.Toolkit.UI/Behaviors/ProgressExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+			}
+		}
+
+		private static IEnumerable<DependencyObject> EnumerateChildren(this DependencyObject reference)
+		{
+			return from x in Enumerable.Range(0, VisualTreeHelper.GetChildrenCount(reference))
+											  select VisualTreeHelper.GetChild(reference, x);
+		}
+	}
+#endif
```
Comment (Xiaoy312) [inline src/Uno.Toolkit.UI/Behaviors/ProgressExtensions.cs:89]:
```suggestion
#endif
	}
```

File: src/Uno.Toolkit.UI/Behaviors/ProgressExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+		}
+
+		public static bool GetIsActive(this FrameworkElement element)
+		{
+			return (bool)element.GetValue(IsActiveProperty);
+		}
+
+#if WINDOWS_UWP || WINDOWS
+		// Copied from Uno.UI.Extensions.ViewExtensions for Windows targets
+		private static IEnumerable<DependencyObject> EnumerateDescendants(this DependencyObject reference)
```
Comment (Xiaoy312) [inline src/Uno.Toolkit.UI/Behaviors/ProgressExtensions.cs:71]: there is no need to redefine these methods
we have our own copy in toolkit
VisualTreeHelperEx:: GetChildren and GetChildren
that are available on all platforms

File: src/Uno.Toolkit.UI/Behaviors/ProgressExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+using Windows.UI.Xaml.Input;
+using ItemsRepeater = Microsoft.UI.Xaml.Controls.ItemsRepeater;
+#endif
+
+
+namespace Uno.Toolkit.UI
+{
+	public static class ProgressExtensions
+	{
+		public static readonly DependencyProperty IsActiveProperty =
```
Comment (Xiaoy312) [inline src/Uno.Toolkit.UI/Behaviors/ProgressExtensions.cs:33]: it is better for you to copy the template like here: 
https://github.com/unoplatform/uno.toolkit.ui/blob/main/src/Uno.Toolkit.UI/Behaviors/DrawerControlBehavior.cs#L35-L49

without setter/getter inside the #region
and the handler outside

Comment (Xiaoy312) [reply src/Uno.Toolkit.UI/Behaviors/ProgressExtensions.cs:33]: could also use a short `<summary>` to describe what it does:
> recursively activate/disactivate all nested progress ring/bar

**Review by Xiaoy312 — COMMENTED** [review]
(comment required)

File: src/Uno.Toolkit.UI/Behaviors/ProgressExtensions.cs [hunk 4]
```diff
@@ -43,36 +54,5 @@ private static void IsActiveChanged(DependencyObject d, DependencyPropertyChange
...
-		}
-
-		public static bool GetIsActive(this FrameworkElement element)
-		{
-			return (bool)element.GetValue(IsActiveProperty);
-		}
-
-#if WINDOWS_UWP || WINDOWS
-		// Copied from Uno.UI.Extensions.ViewExtensions for Windows targets
-		private static IEnumerable<DependencyObject> EnumerateDescendants(this DependencyObject reference)
```
[the remark this answers is not in the document]
Comment (lukeblevins) [reply src/Uno.Toolkit.UI/Behaviors/ProgressExtensions.cs]: @nickrandolph My assumption is that it shouldn't fail on Windows, but it looks like we now conditionally make the file of EnumerateDescendants from Uno.UI internal if on Windows. (as of https://github.com/unoplatform/uno/blame/f8ccae131d86ee9bab216bf10319699569b46a41/src/Uno.UI/Extensions/ViewExtensions.visual-tree.cs#L26)

[the remark this answers is not in the document]
Comment (lukeblevins) [reply src/Uno.Toolkit.UI/Behaviors/ProgressExtensions.cs]: Ended up using VisualTreeHelperEx which @Xiaoy312 pointed out earlier. 👍🏻

File: samples/Uno.Toolkit.Samples/Uno.Toolkit.Samples.Shared/Content/Controls/LoadingViewSample.xaml
Comment (lukeblevins) [reply samples/Uno.Toolkit.Samples/Uno.Toolkit.Samples.Shared/Content/Controls/LoadingViewSample.xaml:1]: Fixed this in my latest commits

File: src/Uno.Toolkit.UI/Behaviors/ProgressExtensions.cs (hunk 2 above)
Comment (lukeblevins) [reply src/Uno.Toolkit.UI/Behaviors/ProgressExtensions.cs]: Addressed with: 08abab3df94b922a6e982d52347a579733d21f43

**Review by kazo0 — APPROVED** [review]

**Review by lukeblevins — APPROVED** [review]

**Comment by kazo0** [comment]
@nickrandolph should this be backported?

**Comment by nickrandolph** [comment]
It can be, but it's new so can go into 5.2 unless needed prior

**Comment by nickrandolph** [comment]
@kazo0 I think given the performance implication this has (ie progress ring continues to run after splash screen), I'll back port it.

**Comment by nickrandolph** [comment]
@mergify backport release/stable/5.1

**Pull request merged**